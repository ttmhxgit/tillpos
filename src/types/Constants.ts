import { ISelect } from "./Select";

export interface ISelectCondition {
  GREATER_THAN: ISelect;
  GREATER_OR_EQUAL_THAN: ISelect;
  EQUAL: ISelect;
  LESS_THAN: ISelect;
  LESS_OR_EQUAL_THAN: ISelect;
}

export interface IProductDiscountType {
  EACH: ISelect;
  ALL: ISelect;
}