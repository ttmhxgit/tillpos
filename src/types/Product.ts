import { ISelect } from "./Select"

export type TProductSize = 'small' | 'medium' | 'large'

export interface IProduct {
    price: number
    name: string
    describe: string
    size: TProductSize
}

export interface IItemAdd {
    id: number;
    item: any
}
export interface IItemProduct extends IItemAdd{
    count: number;
}

export interface IItemsCheckout {
    vendor: string;
    items: Array<IItemProduct>;
}

export interface IDiscountResult {
    totalAmount: number;
    price: number;
    percent: number;
}

export interface IPizzaSelect extends ISelect {
    size?: TProductSize
}