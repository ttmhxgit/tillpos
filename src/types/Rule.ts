export interface IRule {
    itemType: {
        customer: string,
        operator: string,
        amount: number,
        product: string
    },
    discountType: {
        productDiscountType: string,
        amount: number,
        price: number,
        percent: number
    }
}