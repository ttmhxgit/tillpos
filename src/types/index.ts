export * from './Customer'
export * from './Product'
export * from './Rule'
export * from './Select'
export * from './Constants'