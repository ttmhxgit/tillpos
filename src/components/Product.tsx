import React from 'react';
import { IProduct, TCustomer, TCustomerWithQuantity } from '../types'


interface IProductProps {
    product: IProduct,
    onChangeQuantityBtn: (type: 'increase' | 'decrease', index: number) => void,
    index: number
    onChangeQuantityInput: (val: string, index: number) => void,
    customer: TCustomer,
    customerQuanity: TCustomerWithQuantity
}

const Product = ({ product, onChangeQuantityBtn, index, onChangeQuantityInput, customer, customerQuanity }: IProductProps) => {
    return (
        <li className="flex items-center justify-between py-4">
            <div className="flex items-start">
                <div className="ml-4 font-semibold">
                    <p className="text-lg text-secondary">{product.name}</p>

                    <dl className="mt-1 space-y-1 text-xs text-gray-500">
                        <div>
                            <dt className="inline">Description:</dt>
                            &nbsp;
                            <dd className="inline">{product.describe}</dd>
                        </div>

                        <div>
                            <dt className="inline">Price:</dt>
                            &nbsp;
                            <dd className="inline text-base font-semibold">{product.price}</dd>
                        </div>
                    </dl>
                </div>
            </div>

            <div className="flex">
                <button id="decrease" onClick={() => { onChangeQuantityBtn('decrease', index) }} className="w-10 p-1 border-none text-lg rounded-l text-primary font-semibold bg-secondary transition-all hover:bg-secondary hover:text-primary">
                    -
                </button>

                <input
                    id={`input-${index}-${customer}`}
                    onChange={(e) => { onChangeQuantityInput(e.target.value, index)}}
                    value={customerQuanity[customer][product.size] || 0}
                    className="w-10 h-10 p-1 text-lg border-none  text-center bg-secondary text-primary"
                    type="number" min={0}
                />

                <button id="increase" onClick={() => { onChangeQuantityBtn('increase', index) }} className="w-10 p-1 border-none text-lg rounded-r text-primary font-semibold bg-secondary transition-all hover:bg-secondary hover:text-primary">
                    +
                </button>
            </div>
        </li>
    )
}

export default Product;
