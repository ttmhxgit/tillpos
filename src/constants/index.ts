import { IProduct, ICustomer, IRule, ISelectCondition, ISelect, IProductDiscountType, IPizzaSelect, TCustomerWithQuantity } from "../types";
import { toSnakeCase } from '../helpers'

export const Products: IProduct[] = [
    {
        name: "Small Pizza",
        price: 11.99,
        describe: "10'' pizza for one person",
        size: "small",
    }, {
        name: "Medium Pizza",
        price: 15.99,
        describe: "12'' pizza for two persons",
        size: "medium",
    }, {
        name: "Large Pizza",
        price: 21.99,
        describe: "15'' pizza for four persons",
        size: 'large',
    }
]

export const Customers: ICustomer[] = [
    {
        text: "Default",
        value: "default"
    }, {
        text: "Amazon",
        value: "amazon"
    }, {
        text: "Microsoft",
        value: "microsoft"
    }, {
        text: "Facebook",
        value: "facebook"
    },
]

export const DefaultRule: Array<IRule> = [
    {
        itemType: {
            customer: "microsoft",
            operator: "greater_or_equal_than",
            amount: 3,
            product: "small_pizza"
        },
        discountType: {
            productDiscountType: "each",
            amount: 2,
            price: 0,
            percent: 0
        }
    },
    {
        itemType: {
            customer: "amazon",
            operator: "greater_or_equal_than",
            amount: 1,
            product: "large_pizza"
        },
        discountType: {
            productDiscountType: "each",
            amount: 1,
            price: 19.99,
            percent: 0
        }
    },
    {
        itemType: {
            customer: "facebook",
            operator: "greater_or_equal_than",
            amount: 5,
            product: "medium_pizza"
        },
        discountType: {
            productDiscountType: "each",
            amount: 4,
            price: 0,
            percent: 0
        }
    }
];




export const OperatorSelect: ISelectCondition = {
    GREATER_OR_EQUAL_THAN: {
        text: "Greater or equal than",
        value: toSnakeCase("Greater or equal than")
    },
    GREATER_THAN: {
        text: "Greater than",
        value: toSnakeCase("Greater than")
    },
    EQUAL: {
        text: "Equal than",
        value: toSnakeCase("Equal than")
    },
    LESS_THAN: {
        text: "Less than",
        value: toSnakeCase("Less than")
    },
    LESS_OR_EQUAL_THAN: {
        text: "Less or equal than",
        value: toSnakeCase("Less or equal than")
    }
};


export const PizzaSelect: Array<IPizzaSelect> = (function () {
    const data: Array<IPizzaSelect> = Products.map((data) => {
        const formatData: IPizzaSelect = {
            text: data.name,
            value: toSnakeCase(data.name),
            size: data.size,
        };
        return formatData;
    });
    const additionalData: ISelect = {
        text: "Anything",
        value: "anything",
    };
    data.push(additionalData);
    return data;
})();


export const ProductDiscountType: IProductDiscountType = {
    ALL: {
        text: "All",
        value: "all"
    },
    EACH: {
        text: "Each",
        value: "each"
    }
};

export const InitCustomerWithQuanity: TCustomerWithQuantity = {
    amazon: {
        small: 0,
        medium: 0,
        large: 0
    },
    default: {
        small: 0,
        medium: 0,
        large: 0
    },
    microsoft: {
        small: 0,
        medium: 0,
        large: 0
    },
    facebook: {
        small: 0,
        medium: 0,
        large: 0
    },
}