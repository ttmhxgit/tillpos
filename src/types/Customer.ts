import { TProductSize } from "./Product";

export type TCustomer = 'default' | 'amazon' | 'microsoft' | 'facebook';

export interface ICustomer {
    text: string
    value: TCustomer
};


export type TCustomerWithQuantity = {
    [customer in TCustomer]: {
        [product in TProductSize]: number
    }
}