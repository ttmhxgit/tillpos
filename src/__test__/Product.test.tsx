import React from "react";
import Product from "../components/Product";
import Adapter from 'enzyme-adapter-react-16';
import { shallow, ShallowWrapper, configure, mount, ReactWrapper } from "enzyme";
import { TCustomer, TCustomerWithQuantity } from "../types";
import { Products, InitCustomerWithQuanity } from "../constants";


configure({ adapter: new Adapter() });

const testProps = {
    product: Products[0],
    onChangeQuantityBtn: (type: 'increase' | 'decrease', index: number) => { },
    index: 0,
    onChangeQuantityInput: (val: string, index: number) => { },
    customer: 'default' as TCustomer,
    customerQuanity: InitCustomerWithQuanity
}

describe('Checkout', () => {
    let wrapper: ShallowWrapper;
    beforeEach(() => {
        wrapper = shallow(<Product {...testProps} />);
    });

    it("should render correctly", () => {
        expect(wrapper).toMatchSnapshot()
    });


    it("increase 1 item", () => {
        const mockOnchange = jest.fn()
        wrapper = shallow(<Product {...testProps} onChangeQuantityBtn={mockOnchange} />);

        const incButton = wrapper.find("#increase");

        incButton.simulate('click', 'increase', testProps.index);

        wrapper.update();
        expect(mockOnchange).toHaveBeenCalledWith('increase', testProps.index);
    });


    it("decrease 1 item", () => {
        const mockOnchange = jest.fn()
        wrapper = shallow(<Product {...testProps} onChangeQuantityBtn={mockOnchange} />);

        const incButton = wrapper.find("#decrease");

        incButton.simulate('click', 'decrease', testProps.index);

        wrapper.update();
        expect(mockOnchange).toHaveBeenCalledWith('decrease', testProps.index);
    });


    it("change quantity item", () => {
        const mockOnchange = jest.fn()
        wrapper = shallow(<Product {...testProps} onChangeQuantityInput={mockOnchange} />);

        const input = wrapper.find(`#input-${testProps.index}-${testProps.customer}`);

        input.simulate('change', { target: { value: 123 } }, testProps.index)

        wrapper.update();
        expect(mockOnchange).toHaveBeenCalledWith(123, 0);
    });
});
