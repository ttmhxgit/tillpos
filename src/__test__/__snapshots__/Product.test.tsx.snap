// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout should render correctly 1`] = `ShallowWrapper {}`;
